Allow updating of resource policies in compute_disk

chrissng opened the pull request:
<!-- 
Note: You may see "This branch is out-of-date with the base branch"
when you submit a pull request. This is fine! We don't use the GitHub
merge button to merge PRs, and you can safely ignore that message.

Thanks for contributing!
-->

Allow updating of resource policies in compute_disk by using an attachment resource.

TODOs:
- [x] Implement a Terraform resource that attaches a resource policy to a disk.
- [x] Write test that verify that the implementation works as intended.

<!-- CHANGELOG for Downstream PRs.
EXTERNAL CONTRIBUTORS: Your reviewer will most likely fill this in for you, so don't worry about this section!

For some repos (currently Terraform GA/beta providers), we have the
ability to autogenerate CHANGELOGs.

Fill in the following release note code block to have it be added to the CHANGELOG, or leave the block empty if you don't expect this to be added to a downstream PR (i.e. docs-only changes or non-user facing changes)

Please also add any of the following appropriate labels to the PR:
- changelog: bugfix
- changelog: new-resource
- changelog: new-datasource
- changelog: deprecation
- changelog: breaking-change
-->
# Release Note for Downstream PRs (will be copied)
```releasenote
Allow updating of resource policies in compute_disk by using an attachment resource.
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.  They will authorize it to run through our CI pipeline, which will generate downstream PRs.

Thanks for your contribution!  A human will be with you soon.

@rileykarson, please review this PR or find an appropriate assignee.

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.  They will authorize it to run through our CI pipeline, which will generate downstream PRs.

Thanks for your contribution!  A human will be with you soon.

@SirGitsalot, please review this PR or find an appropriate assignee.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
This PR seems not to have generated downstream PRs before, as of fea128574f65ff72de4da0acbafd51622d941534. 
## Pull request statuses
No diff detected in terraform-google-conversion.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/1085 
depends: https://github.com/terraform-providers/terraform-provider-google/pull/4323

rileykarson commented on third_party/terraform/tests/resource_compute_disk_test.go.erb:
```diff
@@ -457,9 +460,64 @@ func TestAccComputeDisk_resourcePolicies(t *testing.T) {
 				ImportState:       true,
 				ImportStateVerify: true,
 			},
+			{
```
If you reuse the block right above here, placing another import step after the config step, the provider will do an import using a blank state and verify that the state is identical to the state used to create the resource. That's sufficient to check that values are set upstream, so you can remove the explicit `Check` by doing so.

Lots of tests in this file use Checks, but we primarily use newer "import-style" tests now 🙂

rileykarson commented on third_party/terraform/tests/resource_compute_disk_test.go.erb:
```diff
@@ -828,4 +886,32 @@ resource "google_compute_disk" "foobar" {
+  snapshot_schedule_policy {
+    schedule {
+      daily_schedule {
+        days_in_cycle = 3
```
Since these values are all that changes, you can templatise them in the original config and supply different values in each stage of the test.

rileykarson commented on third_party/terraform/tests/resource_compute_disk_test.go.erb:
```diff
@@ -457,9 +460,64 @@ func TestAccComputeDisk_resourcePolicies(t *testing.T) {
 				ImportStateVerify: true,
 			},
+			{
+				Config: testAccComputeDisk_resourcePolicies_updated(diskName, policyName2),
```
I think you want to reuse `policyName` here, so that we're updating the resource instead of making a new one?

chrissng commented:
@rileykarson Thank you for the suggestions, I will work on them.

I realise that the update is not fully addressed. According to the Beta API, there is also the [removeResourcePolicies](https://cloud.google.com/compute/docs/reference/rest/beta/disks/removeResourcePolicies) and not just the addResoucePolicies API.

This would suggest that custom update code has to be done on resource_policies property updates that calls both the APIs. It has to handle cases:
1. [Addressed] no resource_policies -> resource_policies X
1. [Not addressed] resource_policies X -> no resource_policies
1. [Not addressed] resource_policies X -> resource_policies Y

For case (1), no need to call removeResourcePolicies. For cases (2) and (3), need to call removeResourcePolicies first before calling addResourcePolicies.

Looking at the magic-modules documentation and codelab, I have not been successful incorporating these behavior using `pre_update` custom code. I am not sure if I am on the right track; do you have any suggestions?

rileykarson commented:
Right, sorry about that. I managed to miss that it was an `addX`/`removeX` subresource originally.
It looks like you've already picked up on the fine-grained resource pattern. It's the best way to support this kind of behaviour with MM right now.

https://github.com/GoogleCloudPlatform/magic-modules/pull/1846/files is a good sample of what adding one looks like. https://github.com/GoogleCloudPlatform/magic-modules/pull/1504/files#diff-fb4f76e7d870258668a3beac48bf164cR276 is also useful as a reference, but it contains some unrelated changes.

rileykarson commented on products/compute/terraform.yaml:
```diff
@@ -302,6 +302,15 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         state as plain-text. [Read more about sensitive data in state](/docs/state/sensitive-data.html).
         Because the API does not return the sensitive key value,
         we cannot confirm or reverse changes to a key outside of Terraform.
+  DiskResourcePoliciesAttachment: !ruby/object:Overrides::Terraform::ResourceOverride
```
You'll also likely need to add an `exclude: true` in `inspec.yaml` and `ansible.yaml` similar to https://github.com/GoogleCloudPlatform/magic-modules/pull/1504/files#diff-63dae33e5de228cd42cb0e6a772a7dbdR22-R23

chrissng commented on products/compute/api.yaml:
```diff
@@ -1275,6 +1275,76 @@ objects:
+    name: 'DiskResourcePoliciesAttachment'
+    kind: 'compute#diskResourcePoliciesAttachment'
+    input: true
+    min_version: beta
```
You are right! I did not realized that it is available in GA. In fact, I found out that resource policies created in terraform returns a selfLink pointing to v1.

chrissng commented on products/compute/terraform.yaml:
```diff
@@ -302,6 +302,15 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         state as plain-text. [Read more about sensitive data in state](/docs/state/sensitive-data.html).
         Because the API does not return the sensitive key value,
         we cannot confirm or reverse changes to a key outside of Terraform.
+  DiskResourcePoliciesAttachment: !ruby/object:Overrides::Terraform::ResourceOverride
```
This is super useful. Thank you for sharing

chrissng commented:
I've implemented your suggestions and the provider works by manually building and testing the tf provider. I wrote a test but I encountered this strange error:
```bash
$ make testacc TEST=./google TESTARGS='-run=TestAccComputeDiskResourcePolicyAttachment_basic'

==> Checking source code against gofmt...
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccComputeDiskResourcePolicyAttachment_basic -timeout 240m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccComputeDiskResourcePolicyAttachment_basic
=== PAUSE TestAccComputeDiskResourcePolicyAttachment_basic
=== CONT  TestAccComputeDiskResourcePolicyAttachment_basic
--- FAIL: TestAccComputeDiskResourcePolicyAttachment_basic (25.04s)
    testing.go:568: Step 1 error: Error reading ComputeDiskResourcePolicyAttachment "v1": googleapi: Error 400: Invalid value 'www.googleapis.com'. Values must match the following regular expression: '(?:(?:[-a-z0-9]{1,63}\.)*(?:[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?):)?(?:[0-9]{1,19}|(?:[a-z0-9](?:[-a-z0-9]{0,61}[a-z0-9])?))', invalidParameter
FAIL
FAIL	github.com/terraform-providers/terraform-provider-google/google	25.079s
GNUmakefile:16: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

I was wondering could it be due to `name` property is not allowed to be a self link?

chrissng commented:
@rileykarson here's the trace log https://gist.github.com/chrissng/428e33d3a76a838ee79579145e09381a

chrissng commented on products/compute/terraform.yaml:
```diff
@@ -302,6 +302,10 @@ overrides: !ruby/object:Overrides::ResourceOverrides
         state as plain-text. [Read more about sensitive data in state](/docs/state/sensitive-data.html).
         Because the API does not return the sensitive key value,
         we cannot confirm or reverse changes to a key outside of Terraform.
+  DiskResourcePolicyAttachment: !ruby/object:Overrides::Terraform::ResourceOverride
```
Recall that the addX and removeX API calls requires the key resourcePolicies that expects a list with a single element. 

That element is populated by {{name}} so it has to be the full URL

chrissng replied:
Works with `projects/xxxx/regions/yyyy/resourcePolicies/zzzz`

chrissng commented:
@rileykarson It looks like we are stuck at troubleshooting the import-style approach to testing here. Should we continue to pursue this or would it be reasonable to review this PR based on the Check approach to testing?

I don't fully understand how the test works, but my guess is that the test does not allow to use full URI in name. If import-style is necessary, I could try to rework the resource a bit - to use `name` simply as the name of the attachment resource, and have an additional property `resource_policy` as the selflink to the resource policy. 

I'm not sure if this is worth the effort. What do you think?

rileykarson commented on products/compute/ansible.yaml:
```diff
@@ -43,6 +45,8 @@ datasources: !ruby/object:Overrides::ResourceOverrides
   RegionDiskType: !ruby/object:Overrides::Ansible::ResourceOverride
     exclude: true
+  ResourcePolicy: !ruby/object:Overrides::Ansible::ResourceOverride
+    exclude: true
```
Can you exclude these resources in the `overrides` list below as well? Sorry, I forgot they were distinct lists.

chrissng commented:
@rileykarson As advised I made some changes:
- specify the ID format
- modify pre_delete and encoder to define the "selfLink" 
- implemented a decoder to ensure read works as intended

Implemented the import-style tests. It seems that the tests creates and destroys the resources successfully, but at the final step I hit an error:
```log
2019/09/04 13:23:58 [DEBUG] Got DONE while polling for operation operation-1567574635200-591b3694a4180-992924ff-ac50b43c's status
2019/09/04 13:23:58 [DEBUG] Finished deleting ResourcePolicy "tf-test-policy-3liopt8v96": map[string]interface {}{"id":"2299310898743014532", "insertTime":"2019-09-03T22:23:55.752-07:00", "kind":"compute#operation", "name":"operation-1567574635200-591b3694a4180-992924ff-ac50b43c", "operationType":"delete", "progress":0, "region":"https://www.googleapis.com/compute/v1/projects/basis-terraform-admin/regions/us-central1", "selfLink":"https://www.googleapis.com/compute/v1/projects/basis-terraform-admin/regions/us-central1/operations/operation-1567574635200-591b3694a4180-992924ff-ac50b43c", "startTime":"2019-09-03T22:23:55.768-07:00", "status":"RUNNING", "targetId":"161969077278143667", "targetLink":"https://www.googleapis.com/compute/v1/projects/basis-terraform-admin/regions/us-central1/resourcePolicies/tf-test-policy-3liopt8v96", "user":"chris@basis-ai.com"}
2019/09/04 13:23:58 [DEBUG] provider has no plugin.Client
2019/09/04 13:23:58 [DEBUG] New state was assigned lineage "f0ce4ad1-9250-a655-25ab-5ccddd36a40e"
2019/09/04 13:23:58 [INFO] terraform: building graph: GraphTypePlanDestroy
2019/09/04 13:23:58 [DEBUG] Starting graph walk: walkPlanDestroy
2019/09/04 13:23:58 [INFO] terraform: building graph: GraphTypeRefresh
2019/09/04 13:23:58 [DEBUG] Starting graph walk: walkRefresh
2019/09/04 13:23:58 [DEBUG] New state was assigned lineage "4c81ae8e-3236-048b-b835-a725d3fd2a18"
2019/09/04 13:23:58 [INFO] terraform: building graph: GraphTypePlanDestroy
2019/09/04 13:23:58 [DEBUG] Starting graph walk: walkPlanDestroy
--- FAIL: TestAccComputeDiskResourcePolicyAttachment_basic (19.57s)
    testing.go:569: Step 1 error: Cannot import non-existent remote object: While attempting to import an existing object to google_compute_disk_resource_policy_attachment.foobar, the provider detected that no object exists with the given id. Only pre-existing objects can be imported; check that the id is correct and that it is associated with the provider's configured region or endpoint, or use "terraform apply" to create a new remote object for this resource.
FAIL
FAIL	github.com/terraform-providers/terraform-provider-google/google	19.591s
GNUmakefile:16: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```
https://gist.github.com/chrissng/428e33d3a76a838ee79579145e09381a#file-debug-attempt3-log

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 573b7836e398cff91df3bee328d5f52e608bf73b. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/192 
depends: https://github.com/terraform-providers/terraform-provider-google/pull/4409 
depends: https://github.com/modular-magician/ansible/pull/382

rileykarson commented on products/compute/api.yaml:
```diff
@@ -1285,6 +1285,67 @@ objects:
+  - !ruby/object:Api::Resource
+    name: 'DiskResourcePolicyAttachment'
+    input: true
+    base_url: projects/{{project}}/zones/{{zone}}/disks
```
```suggestion
    base_url: projects/{{project}}/zones/{{zone}}/disks/{{disk}}
```

We build the import format off the `base_url` parts + `identity` values.

We're constructing [`{{project}}/{{zone}}/{{name}}`](https://github.com/terraform-providers/terraform-provider-google/pull/4409/files#diff-7741ec9e30f8047d04317b3c87911d74R236) right now, but we need `{{disk}}` as well for the import fn to expect the id format of `{{project}}/{{zone}}/{{disk}}/{{name}}`.

I don't think anywhere will use the `base_url`, so this should be a safe change. If you find this breaks something else, we can explicit set the `import_id` beside `id_format`.

chrissng commented:
We're good; the tests passed! 🙏

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 60eb12faf6e79e5548877e8569d2b6fd7fa7f2a7. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/modular-magician/inspec-gcp/pull/200

rileykarson commented on products/compute/api.yaml:
```diff
@@ -1285,6 +1285,67 @@ objects:
+        required: true
+        url_param_only: true
+      - !ruby/object:Api::Type::ResourceRef
+        name: 'zone'
```
In `terraform.yaml` do you mind setting `zone` `required: false` and `default_from_api: true`? The provider-level zone will automatically get inferred that way.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, b41b21e9ef42ef9ee770b5bfbfbac0da2dcfd27e. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

chrissng commented:
@rileykarson I would say this would not have been possible without your help. Thank you for your patience and guidance throughout this PR!

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, b7de2e6fcb07cf5bc3e5d83d411a961d447b0860. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-google-conversion already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/modular-magician/inspec-gcp/pull/207

rileykarson commented on products/compute/api.yaml:
```diff
@@ -1285,6 +1285,67 @@ objects:
+        path: 'name'
+        base_url: 'projects/{{project}}/zones/{{zone}}/operations/{{op_id}}'
+        wait_ms: 1000
+        timeouts: !ruby/object:Api::Timeouts
```
Looks like an x-merge with https://github.com/GoogleCloudPlatform/magic-modules/pull/2283 caused this to generate invalid code, previously this block was ignored. https://github.com/GoogleCloudPlatform/magic-modules/pull/2325 fixes default values, so you can merge that to fix this.

Alternatively, it's probably a bit faster to work around it- you can add `delete_minutes` here as well. I'd define both of these as `4` minutes in that case, we have no strong reason to go lower than the default.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, cd55e96b7964b39e9b8ffa1649c4650d413bf4fb. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
No diff detected in terraform-google-conversion.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I didn't open any new pull requests because of this PR.

rileykarson commented:
This should be merged soon, our generator attached the wrong downstream changes to this PR yesterday. This next run (should) fix it, and it it doesn't I'll reconcile manually.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, b001d149764ff8ea86a540040cf10ac7fad10922. 
## Pull request statuses
terraform-provider-google-beta already has an open PR.
terraform-provider-google already has an open PR.
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/205

rileykarson commented:
Edited above to point to the original conversion PR.

modular-magician commented:
Hi!  I'm the modular magician, I work on Magic Modules.
I see that this PR has already had some downstream PRs generated. Any open downstreams are already updated to your most recent commit, 819ad064646d101db16133c9021be2507c9401fa. 
## Pull request statuses
No diff detected in Ansible.
No diff detected in Inspec.
## New Pull Requests
I built this PR into one or more new PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/1157 
depends: https://github.com/GoogleCloudPlatform/terraform-google-conversion/pull/206 
depends: https://github.com/terraform-providers/terraform-provider-google/pull/4486
